kv/kvserver/closedts/sidetransport: TestRPCConnUnblocksOnStopper failed

kv/kvserver/closedts/sidetransport.TestRPCConnUnblocksOnStopper [failed](https://tanzanite.cluster.engflow.com/invocations/default/1ce408f4-c13b-46c4-ae87-8bab761072c7?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXIvY2xvc2VkdHMvc2lkZXRyYW5zcG9ydDpzaWRldHJhbnNwb3J0X3Rlc3Q=) on master @ [0ea8b622c5354a891cfb867d8d872a0746ba77b3](https://github.com/cockroachdb/cockroach/commits/0ea8b622c5354a891cfb867d8d872a0746ba77b3):


```
=== RUN   TestRPCConnUnblocksOnStopper
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRPCConnUnblocksOnStopper3686204518
    test_log_scope.go:81: use -show-logs to present logs inline
    sender_test.go:554: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport/sender_test.go:554
        	Error:      	Should be true
        	Test:       	TestRPCConnUnblocksOnStopper
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestRPCConnUnblocksOnStopper3686204518
--- FAIL: TestRPCConnUnblocksOnStopper (2.07s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRPCConnUnblocksOnStopper.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39042

--- a/pkg/kv/kvserver/closedts/sidetransport/sender_test.go
+++ b/pkg/kv/kvserver/closedts/sidetransport/sender_test.go
@@ -321,20 +321,18 @@ func TestSenderSameRangeDifferentStores(t *testing.T) {
 
 // mockReceiver is a SideTransportServer.
 type mockReceiver struct {
-	stop chan struct{}
-	mu   struct {
-		syncutil.Mutex
-		called bool
-	}
+	stop     chan struct{}
+	called   atomic.Bool
+	calledCh chan struct{}
 }
 
 var _ ctpb.SideTransportServer = &mockReceiver{}
 
 // PushUpdates is the streaming RPC handler.
 func (s *mockReceiver) PushUpdates(stream ctpb.SideTransport_PushUpdatesServer) error {
-	s.mu.Lock()
-	s.mu.called = true
-	s.mu.Unlock()
+	if s.called.CompareAndSwap(false, true) {
+		close(s.calledCh)
+	}
 	// Block the RPC until close() is called.
 	<-s.stop
 	return nil
@@ -342,14 +340,9 @@ func (s *mockReceiver) PushUpdates(stream ctpb.SideTransport_PushUpdatesServer) 
 
 func newMockReceiver() *mockReceiver {
 	return &mockReceiver{
-		stop: make(chan struct{}),
+		stop:     make(chan struct{}),
+		calledCh: make(chan struct{}),
 	}
-}
-
-func (s *mockReceiver) getCalled() bool {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	return s.mu.called
 }
 
 // sideTransportGRPCServer wraps a Receiver (a real one of a mock) in a gRPC
@@ -521,6 +514,9 @@ func TestRPCConnUnblocksOnStopper(t *testing.T) {
 
 	s.publish(ctx)
 	require.Len(t, s.connsMu.conns, 1)
+	// Wait until at least one update has been delivered. This means the rpcConn
+	// task has been started.
+	<-srv.mockReceiver().calledCh
 
 	// Now get the rpcConn to keep sending messages by calling s.publish()
 	// repeatedly. We'll detect when the rpcConn is blocked (because the Receiver
@@ -550,8 +546,6 @@ func TestRPCConnUnblocksOnStopper(t *testing.T) {
 	// Stop the stopper. If this doesn't timeout, then the rpcConn's task must
 	// have been unblocked.
 	stopper.Stop(ctx)
-
-	require.True(t, srv.mockReceiver().getCalled())
 }
 
 // Test a Sender and Receiver talking gRPC to each other.

